sw/device/tests/dif/dif_gpio_test.cc

```diff
@@ -0,0 +1,420 @@
+}
+
+// Output mode tests
+class OutputModeTest : public DifGpioTestWithDevice {};
+
+TEST_F(OutputModeTest, NullArgs) {
+  EXPECT_EQ(dif_gpio_output_mode_all_set(nullptr, 0xFFFFFFFF),
+            kDifGpioResultInvalidArgument);
+  EXPECT_EQ(dif_gpio_output_mode_pin_set(nullptr, 0, true),
+            kDifGpioResultInvalidArgument);
+  EXPECT_EQ(dif_gpio_output_mode_masked_set(nullptr, 0xFFFFFFFF, 0xFFFFFFFF),
+            kDifGpioResultInvalidArgument);
+}
+
+TEST_F(OutputModeTest, AllPins) {
+  uint32_t const val = 0xA5A5A5A5;
+  EXPECT_WRITE32(GPIO_DIRECT_OE_REG_OFFSET, val);
+
+  EXPECT_EQ(dif_gpio_output_mode_all_set(&gpio_, val), kDifGpioResultOK);
+}
+
+TEST_F(OutputModeTest, SinglePin) {
+  EXPECT_WRITE32(GPIO_MASKED_OE_LOWER_REG_OFFSET, {{16, 1}, {0, 1}});
+  EXPECT_EQ(dif_gpio_output_mode_pin_set(&gpio_, 0, true), kDifGpioResultOK);
+
+  EXPECT_WRITE32(GPIO_MASKED_OE_LOWER_REG_OFFSET, {{31, 1}, {15, 0}});
+  EXPECT_EQ(dif_gpio_output_mode_pin_set(&gpio_, 15, false), kDifGpioResultOK);
+
+  EXPECT_WRITE32(GPIO_MASKED_OE_UPPER_REG_OFFSET, {{16, 1}, {0, 1}});
+  EXPECT_EQ(dif_gpio_output_mode_pin_set(&gpio_, 16, true), kDifGpioResultOK);
+
+  EXPECT_WRITE32(GPIO_MASKED_OE_UPPER_REG_OFFSET, {{31, 1}, {15, 0}});
+  EXPECT_EQ(dif_gpio_output_mode_pin_set(&gpio_, 31, false), kDifGpioResultOK);
+}
+
+TEST_F(OutputModeTest, Masked) {
+  EXPECT_WRITE32(GPIO_MASKED_OE_LOWER_REG_OFFSET, 0xCDCD3322);
+  EXPECT_WRITE32(GPIO_MASKED_OE_UPPER_REG_OFFSET, 0xABAB5544);
+  EXPECT_EQ(dif_gpio_output_mode_masked_set(&gpio_, 0xABABCDCD, 0x55443322),
+            kDifGpioResultOK);
+
+  EXPECT_WRITE32(GPIO_MASKED_OE_LOWER_REG_OFFSET, 0xCDCD3322);
+  EXPECT_EQ(dif_gpio_output_mode_masked_set(&gpio_, 0x0000CDCD, 0x55443322),
+            kDifGpioResultOK);
+
+  EXPECT_WRITE32(GPIO_MASKED_OE_UPPER_REG_OFFSET, 0xABAB5544);
+  EXPECT_EQ(dif_gpio_output_mode_masked_set(&gpio_, 0xABAB0000, 0x55443322),
+            kDifGpioResultOK);
+}
+
+// Input noise filter tests
+class InputFilterTest : public DifGpioTestWithDevice {};
+
+TEST_F(InputFilterTest, NullArgs) {
+  EXPECT_EQ(dif_gpio_input_noise_filter_masked_enable(nullptr, 0xFFFFFFFF),
+            kDifGpioResultInvalidArgument);
+  EXPECT_EQ(dif_gpio_input_noise_filter_masked_disable(nullptr, 0xFFFFFFFF),
+            kDifGpioResultInvalidArgument);
+}
+
+TEST_F(InputFilterTest, MaskedEnable) {
+  const uint32_t val = 0xABABABAB;
+  EXPECT_READ32(GPIO_CTRL_EN_INPUT_FILTER_REG_OFFSET, 0x0);
+  EXPECT_WRITE32(GPIO_CTRL_EN_INPUT_FILTER_REG_OFFSET, val);
+
+  EXPECT_EQ(dif_gpio_input_noise_filter_masked_enable(&gpio_, val),
+            kDifGpioResultOK);
+}
+
+TEST_F(InputFilterTest, MaskedDisable) {
+  const uint32_t val = 0xABABABAB;
+  EXPECT_READ32(GPIO_CTRL_EN_INPUT_FILTER_REG_OFFSET, 0xFFFFFFFF);
+  EXPECT_WRITE32(GPIO_CTRL_EN_INPUT_FILTER_REG_OFFSET, ~val);
+
+  EXPECT_EQ(dif_gpio_input_noise_filter_masked_disable(&gpio_, val),
+            kDifGpioResultOK);
+}
+
+// IRQ tests
```

Comment:
Nit: I don't think this comment adds anything.

Reply:
Removed.